Async stack trace structure (from dev)

clementbera opened the pull request:
Version rebased on dev branch for SMarr review.

So far basic synchronous on asynchronous benchmark are working with the different settings enabled. With the settings disabled everything should keep working too. Still need to work on printing the stack with the different settings enabled or not.

Three additional settings:
--async_stack_traces_structure: enables the shadow stack entry, allowing to recreate the stack across asynchronous calls.
--async_stack_traces_method_cache: enables per method cache to CachedDispatchNode or LexicallyBoundDispatchNode, avoiding to create shadow stack entry if a method is called from a single dispatch node.
--async_stack_traces_inline_cache: enables per shadow stack entry creation site cache, avoiding to create repeatedly the same shadow stack entry in loops, the first few times a loop is created.

async_stack_traces_structure has to be enabled for the other ones to change anything. The other 2 settings can be enabled separately or together.

clementbera committed:
Rebased AsyncStackTraceStructureBranch on dev. Made it work with async stack trace structure disabled

clementbera committed:
added all changes

clementbera committed:
fix bug leading to async stack trace working with tinyBench again (enabled)

clementbera committed:
finally asynchronous send working. Now need to ensure it works correctly and look at problematic cases

clementbera committed:
Added comments and fixed exception handling which was loosing ShadowStackEntry

clementbera committed:
fixed class access and promise pair creation cases where shadow stack entry was broken

clementbera committed:
re-enabled SMarr assertion since now top SSEntry is created explicitely

clementbera committed:
split StackIterator in its own file

clementbera committed:
fix bug preventing to use the code with Async stack structure enabled. Started to debug printing of async stack trace

clementbera committed:
resolve conflicts

clementbera committed:
made printStackTrace work again with SSEntry and with SSEntry but no method cache

clementbera committed:
removed debugging code. Added support for async stack trace printing with method cache

smarr commented on build.xml:
```diff
@@ -21,11 +21,11 @@
     <property name="truffle.build" location="${truffle.dir}/mxbuild/dists" />
     <property name="somns-deps.version" value="0.3.6" />
-    <property name="checkstyle.version" value="8.11" />
+    <property name="checkstyle_checkse.version" value="8.11" />
```
Why this change?

clementbera replied:
I think this was generated by mx.

smarr replied:
No, please revert. mx doesn't know anything about our build system.

smarr commented on docs/basic-setup.md:
````diff
@@ -48,4 +48,4 @@ Afterwards, the simple Hello World program is executed with:
 ```bash
 ./som core-lib/Hello.ns
-```
+```
````
Strange change?

clementbera replied:
Yes, the change should be removed but I don't know how. I checked out from dev and it is still there.

smarr commented on src/som/interpreter/Invokable.java:
```diff
@@ -46,6 +46,10 @@ public final boolean isAtomic() {
     return isAtomic;
   }
 
+  public ExpressionNode getRoot() {
```
`Node.getRootNode()` isn't any good for this?

clementbera replied:
I think Node.getRootNode answers the root node, i.e. the method or primitive, while this method  answers the sequence of expression held by the root node. So I would say no.

smarr replied:
would `getBodyNode()` be an appropriate name?

clementbera replied:
Done

smarr commented on src/som/interpreter/Method.java:
```diff
@@ -77,6 +80,20 @@ public boolean equals(final Object o) {
     return false;
   }
 
+  public void setNewCaller(final BackCacheCallNode caller) {
```
You are changing a `@CompilationFinal` field here, do you need to invalidate any code that may have been compiled with this assumption backed in?

clementbera replied:
Oh, I thought @CompilationFinal was taking care of that for me. Yes code needs to be invalidated.

clementbera replied:
So I removed the @CompilationFinal of this variable. The problem is in BackCacheCallNode >> multipleCaller (). When this method is called, all the optimized code using the BackCacheCallNode has to be invalidated. I am not sure on how to do that.

smarr replied:
Not sure this is the ideal example: https://github.com/smarr/SOMns/blob/dev/src/som/interpreter/objectstorage/ObjectLayout.java#L153

Assumptions are the way to invalidate code "non-locally". Any code that is compiled checking an assumption is registered for being discarded when the assumption is invalidated. In the example, there is an assumption for the memory layout of classes. It is invalidated, when the class gets a new layout assigned (which in return is going to be used by its instances).

smarr replied:
Not sure why I don't see these in the review, but we probably still want this to be `@CompilationFinal` for performance reasons.

clementbera replied:
I tried to fix the concern with invalidation (the 4 comments) in my latest commit.

Since the field uniqueCaller is used in Method only during relinking, it is not compiled by Graal and does not have to be compilation final. The fields uniqueCaller in CachedDispatchNode and LexicallyBoundDispatchNode have however to be. I created an assumption at creation time which is invalidated when the method makeMultipleCaller (was: multipleCaller) is executed.

Looks like some of the test code I have still runs that way.

smarr commented on src/som/interpreter/actors/ErrorPromiseNode.java:
```diff
@@ -10,21 +10,23 @@
 @GenerateNodeFactory
-@Primitive(primitive = "actorsError:with:isBPResolver:isBPResolution:")
+@Primitive(primitive = "actorsError:with:entry:isBPResolver:isBPResolution:")
+
```
Why the space?
Also, you seem to change the signature of the specialization, I'd assume that primitive signature would need to change then, too?

clementbera replied:
I removed the space.

I am not sure about what you are asking for the signature. The signature of AbstractPromiseResolutionNode was also changed, is that what you mean?

smarr reviewed (commented):
Some very initial comments, only skimming over the first bits

clementbera committed:
removed space

clementbera committed:
invalidate logic

smarr commented:
@clementbera thanks for all the work so far. Before I continue trying to understand all the changes, what's the status with the tests? Any estimate on whether this is a fundamental issue or just about working through the various scattered small issues here and there?

smarr commented on src/som/interpreter/SArguments.java:
```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
+    }
+  }
+
+  public static Object[] getPlain1ArgumentWithReceiver(final Object receiver,
```
In Java, we got method overloading for better or for worse. So, I don't think we need to encode arity in method names here.

clementbera replied:
So I changed to getPlainXArgumentWithReceiver with variable number of remaining parameters and made a single function instead of 3.

smarr commented on src/som/interpreter/SArguments.java:
```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
+
+  public static ShadowStackEntry instantiateShadowStackEntry(final ShadowStackEntry previous,
+      final Node expr, final boolean async) {
+    if (async) {
```
Could you add the following? This is a compile time check that async is known and can be used for optimization. It's not a hint that this is the case, it will only result in an error if it isn't a constant value. Good for debugging of performance issues.

```suggestion
    CompilerAsserts.partialEvaluationConstant(async);
    if (async) {
```

clementbera replied:
Done, it works.

smarr commented on src/som/interpreter/SArguments.java:
```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
+
+  public static void setShadowStackEntry(final Object[] args,
+      final ShadowStackEntry entry) {
+    // assert args[args.length - 1] == null : "Assume shadow stack entry is not already set.";
```
Hm, why is this commented out?
Seems like a sensible assertion, or does it not hold? If so, in which case does it fail, should the assertion be removed or documented?

clementbera replied:
It should be uncommented. Done.

smarr replied:
@ctrlpz please do not commit such a change. These assertions are there for a reason. It just cost me a lot of time that the assertion wasn't there.

smarr commented on src/som/interpreter/nodes/dispatch/BackCacheCallNode.java:
```diff
@@ -0,0 +1,47 @@
+        || !VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE;
+  }
+
+  void uniqueCaller();
```
This and the next method seem to be mutating methods. Perhaps to be clear `set` or `make` would be a good prefix for them?

smarr commented on src/som/interpreter/nodes/dispatch/CachedDispatchNode.java:
```diff
@@ -2,19 +2,30 @@
-  @Child private AbstractDispatchNode nextInCache;
+  @Child private DirectCallNode         cachedMethod;
+  @Child private AbstractDispatchNode   nextInCache;
+  @CompilationFinal private boolean     uniqueCaller;
```
Same issue as in other node, you'll need an assumption to make this correct.

smarr commented on src/som/interpreter/nodes/dispatch/LexicallyBoundDispatchNode.java:
```diff
@@ -1,22 +1,33 @@
 
-  @Child private DirectCallNode cachedMethod;
+  @Child private DirectCallNode         cachedMethod;
+  @CompilationFinal private boolean     uniqueCaller;
```
And all the other nodes: needs assumption

clementbera committed:
removed incorrect changes

clementbera committed:
resolving SMarr comment

clementbera committed:
solving SMarr comments

clementbera committed:
fixing SMarr comments;

clementbera committed:
fixed SMarr comments

clementbera committed:
tried to add Assumptions to fix SMarr concern about invalidation of node

clementbera committed:
Fixed assumptions with SMarr

ctrlpz committed:
Fixes to see the trace when an error is thrown

- commented assertion in SArguments in order to avoid an assertion error
- add code to print the stack trace when an error is thrown in a SOMns program

ctrlpz committed:
Fix compilation errors

- change private attributes to protected in LexicallyBoundDispatchNode
- call LexicallyBoundDispatchNode Gen.create() in SInvokable because LexicallyBoundDispatchNode cannot be instantiate

ctrlpz committed:
Fix in CachedDispatchNode class (by Clement)

clementbera commented:
Before we added our change, there was already a try catch getting invalid assumption and replacing the node by the next in cache. This happened when guard.entryMatches raised the invalid assumption exception.

With new code this led to an infinite loop since the node was replaced with another node with the exact same guard.

I added a branch in the catch to check if the problem came from the guard or the stillUniqueCaller assumption. Depending on where it comes from, the problem is dealt with differently. In both cases, it leads in the end to a stable ast with a fast path.

smarr commented:
@clementbera ok, thanks for the explanation!

smarr committed:
Pass VirtualFrame for assertions that may need it

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Pass null instead of frame for some things that should not fail

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Fix formatting

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make shadow stack creation conditional on VmSettings

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Reenable disabled assertion

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Read operations need frame for constructors

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Pass null to read of serialization operations for now

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Set shadow stack for basic interpreter tests, too

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
In ReceiverMessage, set shadow stack only when enabled

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add assertions

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Make ShadowStackEntryLoad classes private and use factory method instead

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
This PR is out of date.

A new one may be opened based on the current branch derived from Carmen's version, which is based on Clement's code: https://github.com/smarr/SOMns/tree/ct/async-stack-traces
